src/host/broadcast.jl

```diff
@@ -112,12 +111,12 @@ function Base.map(f, x::AnyGPUArray, xs::AbstractArray...)
         # see `broadcast`
         ElType = BrokenBroadcast{ElType}
     end
-    dest = similar(x, ElType, common_length)
+    dest = similar(first(xs), ElType, common_length)
 
     return map!(f, dest, xs...)
 end
 
-function Base.map!(f, dest::AnyGPUArray, xs::AbstractArray...)
+function Base.map!(f, dest::AnyGPUArray, xs::AnyGPUArray...)
```

Comment:
Hi @maleadt, @simone-silvestri and I are wondering what motivates this change in how `map!` is dispatched? Previously we were able to call `map!(f, dest::AnyGPUArray, custom_array_on_gpu)` and things would "just work" for appropriate `custom_array_on_gpu`.

With this change we have to figure out how to ensure our custom arrays subtype `AnyGPUArray`. Which may be possible (in versions past, `AnyGPUArray` was a bit more specific than it seems to be now). It's a fair amount of work it looks like for us to make this change so that we can use `map!`, but may be worth it. But just wanted to check here first before getting into that.